### PR TITLE
Implement dplyr API

### PR DESCRIPTION
**Pull request by alexkyllo**
Implement dplyr API

### CODE REVIEW

**Comment by alexkyllo** [comment]
Note that there are still a few unsupported dplyr verbs (pull, transmute, do, top_n, select_if…). If a verb doesn't have a test, you can assume it's not implemented yet.

Also note I have yet to implement R-style naming of duplicate columns after a join (e.g. Species.x, Species.y) and instead it will do the Kusto-style naming (e.g. Species, Species1).

**Comment by Hong-Revo** [comment]
Running the tests resulted in 3 warnings:

```
test-translate.r:67: warning: filter errors on missing symbols
'glue::collapse' is deprecated.
Use 'glue_collapse' instead.
See help("Deprecated") and help("glue-deprecated").

test-translate.r:265: warning: inner_join() on two columns translates correctly
the condition has length > 1 and only the first element will be used

test-translate.r:287: warning: inner_join() on two differently named columns translates correctly
the condition has length > 1 and only the first element will be used
```
Should also add  glue to the DESCRIPTION file.

**Comment by Hong-Revo** [comment]
> Note that there are still a few unsupported dplyr verbs (pull, transmute, do, top_n, select_if…). 

`transmute` is basically a mutate followed by a select, so that one should be easy. Or at least it is in SQL, where you can assume the query optimizer will fix things up. Might be trickier with Kusto since it's less abstract.

**Comment by Hong-Revo** [comment]
Just noticed you added a remotes: r-dbi/DBI to the DESCRIPTION. Is there anything in dev DBI that we need?

**Comment by Hong-Revo** [comment]
Also need to remove AzureRMR as an imported namespace, since `AzureRMR::is_empty` clashes with `rlang::is_empty` (although they do the same thing). This shouldn't break anything since I actually qualified all AzureRMR references anyway, oops....

**Comment by alexkyllo** [comment]
I don't have a specific reason for getting DBI from github over CRAN. I should just remove the DBI stuff for now. I only put in the boilerplate class and methods with no definition, and then I didn't get around to implementing it.

**Comment by alexkyllo** [comment]
Ok, removed the DBI stuff for now, as it's not really part of this feature. We can add it back later. Removed the import(AzureRMR) to clear that warning.

I'm not totally sure what the other warnings mean, but I'll look into them. The glue::collapse deprecation warning, I think is bubbling up from a dependency because I'm not calling it anywhere directly, AFAIK.

**Comment by Hong-Revo** [comment]
As a general style guide, I don't put spaces around `=` in named function arguments, eg `f(a=42)` instead of `f(a = 42)`. I do put spaces around other operators, eg `a + b` instead of `a+b` and `x <- y` instead of `x<-y` or `x=y`.

File: DESCRIPTION [hunk 1]
```diff
@@ -14,7 +14,6 @@ VignetteBuilder: knitr
 Depends:
     R (>= 3.3)
 Imports:
```
[the remark this answers is not in the document]
Comment (alexkyllo) [reply DESCRIPTION]: Added rlang in [f987b8e](https://github.com/cloudyr/AzureKusto/pull/6/commits/f987b8e8cfe7a6f0ab12051d67569d1953ad4b3a) . glue is only a transitive dependency.

File: R/kql-build.R [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+        }
+    }
+    else
+    {
+        has_agg <- FALSE
+        verb <- "extend "
+        by <- ""
+    }
+
+    stmts <- mapply(translate_kql, assigned_exprs)
```
[the remark this answers is not in the document]
Comment (alexkyllo) [reply R/kql-build.R]: Not sure I understand--it seems to be working as intended as far as I can tell, meanwhile adding `simplify=FALSE` causes a test to fail.

File: R/kql-render.R [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+{
+    tblname <- sprintf("database('%s').%s", q$src$database, escape(ident(q$src$table)))
+    q_str <- paste(unlist(q$ops[-1]), collapse = "\n| ")
+    if (nchar(q_str) > 0) {
+        q_str <- kql(paste(tblname, q_str, sep="\n| "))
+    } else {
+        q_str <- tblname
+    }
+
+    cat(q_str, "\n")
```
[the remark this answers is not in the document]
Comment (alexkyllo) [reply R/kql-render.R]: Yep, removing the cat() call.

**Comment by alexkyllo** [comment]
Warnings are gone now :+1:

File: R/kql-build.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hong-Revo) [reply R/kql-build.R]: `mapply`, like `sapply`, attempts to simplify the result from a list into a matrix/vector if possible. If this is the behavior you're after, that's all good. Just be sure you also take into account the scenario where the simplification doesn't happen, in which case you'll have to deal with a list.

File: R/kql-escape.R [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+{
+    x <- as.character(x)
+    escape.character(x, parens = parens, collapse = collapse)
+}
+
+#' @export
+escape.Date <- function(x, parens = NA, collapse = ", ")
+{
+    x <- as.character(x)
+    escape.character(x, parens = parens, collapse = collapse)
```
Comment (Hong-Revo) [inline R/kql-escape.R]: can simplify to `escape(as.character(x), parens=parens, collapse=collapse)`

File: R/tbl.R [hunk 5]
```diff
@@ -1,39 +1,211 @@
...
+#'
+#' @keywords internal
+#' @export
+#' @examples
+#' library(dplyr)
+#' df <- data.frame(x = 1, y = 2)
+#'
+#' df <- tbl_abstract(df, src = simulate_kusto())
+#' df %>% summarise(x = sd(x)) %>% show_query()
+tbl_abstract <- function(df, table_name, src = NULL) {
```
Comment (Hong-Revo) [inline R/tbl.R]: Consider renaming the class to `tbl_kusto_abstract`, current name is very generic and may clash with other packages

**Comment by Hong-Revo** [comment]
Awesome job! Just had a couple more suggestions, other than that I think we can pretty much merge this.

**Comment by alexkyllo** [comment]
Made both those changes. Also added `Encoding: UTF-8` to DESCRIPTION to silence a warning from roxygen2.

**Comment by alexkyllo** [comment]
QQ: Should `collect()` return a data.frame, a tibble, or have an optional argument to choose between the two? I was calling `as.tibble` but I removed it because I'm not sure if you want to take a dependency on the tibble pkg.

**Comment by Hong-Revo** [comment]
> QQ: Should collect() return a data.frame, a tibble, or have an optional argument to choose between the two? I was calling as.tibble but I removed it because I'm not sure if you want to take a dependency on the tibble pkg.

Returning a tibble is fine, because we're in the tidyverse now. By requiring dplyr we've also required everything it depends on.

**Comment by Hong-Revo** [comment]
Note that you don't actually need to import tibble, as dplyr re-exports a lot of its stuff. You should be able to call `dplyr::as_data_frame`, for instance.

**Comment by alexkyllo** [comment]
Ok, calling `dplyr::as_tibble()` (without the qualification since it's already imported), instead of `tibble::as_tibble`

**Pull request merged**